compiler/rustc_target/src/asm/arm.rs

```diff
@@ -88,6 +88,18 @@ fn frame_pointer_r7(
     }
 }
 
+fn rwpi_enabled(
+    _arch: InlineAsmArch,
+    _has_feature: impl FnMut(&str) -> bool,
+    target: &Target,
+) -> Result<(), &'static str> {
+    if let RelocModel::Rwpi | RelocModel::RopiRwpi = target.relocation_model {
```

Comment:
Should we only allow r9 when the relocation model was explicitly given to allow for more freedom in choosing which relocation model will be used by default without risking breaking this?

Reply:
That might be a good idea, though it would make the diagnostic less clear if the user hasn't specifically opted in to that mode. I wanted to add a "note" to it, but this is my first time contributing to rustc and I don't know how to do that, much less how to make it work in the context of `def_regs!`.

How does one check if a target attribute is explicitly set?

Reply:
You can check the value of `sess.opts.cg.relocation_model`. By the way I think `target.relocation_model` is only the default relocation model in all cases and not the one that will actually be used when compiling if `-Crelocation-model` is used.

Reply:
Good catch, I didn't think of that since in my use case I'm using a custom target. I can fix that up. I'd still like to improve the diagnostics though.

Reply:
Oh, `Session` already has a `relocation_model` getter that does the logic I was going to write anyway.

Reply:
Ah right, I can't actually do that because the asm parser isn't passed the `Session`, only the `Target`... Looks like I would need to change the definition of `parse` in the entire `rustc_target::asm` module at least, including arch-specific filter functions, then alter `rustc_ast_lowering::LoweringContext` to pass `Session` as well. That feels like a fairly invasive change just to make a small thing more correct, but the overall effect should be negligible. If the maintainers are okay with it, I can go ahead and do it.

Reply:
This won't work since `rustc_target` can't depend on `rustc_session` since that would create a circular dependency.